Wrappers/Python/cil/optimisation/functions/SVRGFunction.py

```diff
@@ -0,0 +1,248 @@
+        if (np.isinf(self.update_frequency) == False and (self._svrg_iter_number % (self.update_frequency)) == 0):
+
+            return self._update_full_gradient_and_return(x, out=out)
+
+        else:
+
+            self.function_num = self.sampler.next()
+
+            
+            
+            if self.function_num >= self.num_functions or self.function_num<0 :
+                raise IndexError('The sampler has outputted an index larger than the number of functions to sample from. Please ensure your sampler samples from {0,1,...,len(functions)-1} only.')
+            return self.approximate_gradient(x, self.function_num, out=out)
+
+        
+
+    def approximate_gradient(self, x, function_num, out=None):
+        """ Calculates the gradient of the selected function, indexed by `function_number` in {0,...,len(functions)-1}, at the point :math:`x` and at the stored snapshot :math:`\tilde{x}` before returning the stochastic gradient
+            .. math ::
+                \nabla f_i(x_k) - \nabla f_i(\tilde{x}) + \nabla \sum_{i=0}^{n-1}f_i(\tilde{x})
+        
+        Parameters
+        ----------
+        x : DataContainer ( e.g. ImageData)
+        out: return DataContainer, if `None` a new DataContainer is returned, default `None`.
+        function_num: `int` 
+            Between 0 and the number of functions in the list  
+        Returns
+        --------
+        DataContainer (e.g. ImageData)
+            the value of the approximate gradient of the sum function at :code:`x` given a `function_number` in {0,...,len(functions)-1}
+        """
+    
+
+        self._svrg_iter_number += 1
+
+        self.stoch_grad_at_iterate = self.functions[function_num].gradient(x)
+
+        if self.store_gradients is True:
+            self._stochastic_grad_difference = self.stoch_grad_at_iterate.sapyb(
+                1., self._list_stored_gradients[function_num], -1.)
+        else:
+            self._stochastic_grad_difference = self.stoch_grad_at_iterate.sapyb(
+                1., self.functions[function_num].gradient(self.snapshot), -1.)
+
+        self._update_data_passes_indices([function_num])
+
+        if out is None:
+            out = self._stochastic_grad_difference.sapyb(
+                self.num_functions, self._full_gradient_at_snapshot, 1.)
+        else:
+            self._stochastic_grad_difference.sapyb(
+                self.num_functions, self._full_gradient_at_snapshot, 1., out=out)
+
+        return out
+
+    def _update_full_gradient_and_return(self, x, out=None):
+        """
+        Takes a "snapshot" at the point :math:`x`, saving both the point :math:` \tilde{x}=x` and its gradient :math:`\sum_{i=0}^{n-1}f_i{\tilde{x}}`. The function returns :math:`\sum_{i=0}^{n-1}f_i{\tilde{x}}` as the gradient calculation. If :code:`store_gradients==True`, the gradient of all the :math:`f_i`s is computed and stored at the "snapshot"..
```

Comment:
```suggestion
        Takes a "snapshot" at the point :math:`x`. The function returns :math:`\sum_{i=0}^{n-1}f_i{\tilde{x}}` as the gradient calculation. If :code:`store_gradients==True`, the gradient of all the :math:`f_i`s is stored, otherwise only  the sum of the gradients and the snapshot point :math:` \tilde{x}=x` are stored.
```